File: wasm/src/trc/frame.zig
const std = @import("std");
const trace_dlc = @import("../trace/dlc.zig");
const trace_frame = @import("../trace/frame.zig");
const trace_time = @import("../trace/time.zig");

pub const Version = enum {
    v10,
    v11,
    v12,
    v13,
    v20,
    v21,
    v30,

    pub fn fromText(text: []const u8) !Version {
        if (std.mem.eql(u8, text, "1.0")) return .v10;
        if (std.mem.eql(u8, text, "1.1")) return .v11;
        if (std.mem.eql(u8, text, "1.2")) return .v12;
        if (std.mem.eql(u8, text, "1.3")) return .v13;
        if (std.mem.eql(u8, text, "2.0")) return .v20;
        if (std.mem.eql(u8, text, "2.1")) return .v21;
        if (std.mem.eql(u8, text, "3.0")) return .v30;
        return error.UnsupportedTrcVersion;
    }

    pub fn isV2(self: Version) bool {
        return self == .v20 or self == .v21;
    }
};

pub const Kind = trace_frame.Kind;
pub const Id = trace_frame.Id;
pub const Frame = trace_frame.Frame;

pub fn idFromText(text: []const u8) !Id {
    if (text.len != 4 and text.len != 8) return error.InvalidId;
    const value = try std.fmt.parseUnsigned(u32, text, 16);
    if (text.len == 8) {
        if (value > 0x1fff_ffff) return error.InvalidId;
        return Id.extended(value);
    }
    if (value > 0x7ff) return error.InvalidId;
    return Id.standard(@intCast(value));
}

pub const ColumnMap = struct {
    number: ?usize = null,
    offset: ?usize = null,
    record_type: ?usize = null,
    bus: ?usize = null,
    id: ?usize = null,
    direction: ?usize = null,
    reserved: ?usize = null,
    data_len: ?usize = null,
    dlc: ?usize = null,
    data: ?usize = null,
    token_count_before_data: usize = 0,

    pub fn fromText(text: []const u8) !ColumnMap {
        var map: ColumnMap = .{};
        var index: usize = 0;
        var parts = std.mem.splitScalar(u8, text, ',');
        while (parts.next()) |raw_part| : (index += 1) {
            const part = std.mem.trim(u8, raw_part, " \t\r");
            if (std.mem.eql(u8, part, "N")) {
                map.number = index;
            } else if (std.mem.eql(u8, part, "O")) {
                map.offset = index;
            } else if (std.mem.eql(u8, part, "T")) {
                map.record_type = index;
            } else if (std.mem.eql(u8, part, "B")) {
                map.bus = index;
            } else if (std.mem.eql(u8, part, "I")) {
                map.id = index;
            } else if (std.mem.eql(u8, part, "d")) {
                map.direction = index;
            } else if (std.mem.eql(u8, part, "R")) {
                map.reserved = index;
            } else if (std.mem.eql(u8, part, "l")) {
                map.data_len = index;
            } else if (std.mem.eql(u8, part, "L")) {
                map.dlc = index;
            } else if (std.mem.eql(u8, part, "D")) {
                map.data = index;
            } else {
                return error.UnsupportedTrcColumn;
            }
        }

        if (map.offset == null or map.record_type == null or map.id == null or
            map.direction == null or map.data == null or (map.data_len == null and map.dlc == null))
        {
            return error.InvalidTrcColumns;
        }
        map.token_count_before_data = map.data.?;
        return map;
    }
};

pub fn parseTimestampMsToNs(text: []const u8) !u64 {
    if (text.len == 0 or text[0] == '-') return error.InvalidTimestamp;

    var parts = std.mem.splitScalar(u8, text, '.');
    const ms_text = parts.next() orelse return error.InvalidTimestamp;
    const fraction_text = parts.next();
    if (parts.next() != null) return error.InvalidTimestamp;

    const milliseconds = try std.fmt.parseUnsigned(u64, ms_text, 10);
    var ns = try std.math.mul(u64, milliseconds, std.time.ns_per_ms);

    if (fraction_text) |fraction| {
        const fraction_ns = try trace_time.decimalFractionToUnits(fraction, std.time.ns_per_ms, 6, .reject);
        ns = try std.math.add(u64, ns, fraction_ns);
    }

    return ns;
}

pub fn parseDlc(text: []const u8) !u8 {
    const dlc = try std.fmt.parseUnsigned(u8, text, 10);
    if (dlc > 15) return error.InvalidDlc;
    return dlc;
}

pub fn parsePayloadLength(text: []const u8) !u8 {
    const payload_len = try std.fmt.parseUnsigned(u8, text, 10);
    if (payload_len > 64) return error.InvalidPayloadLength;
    return payload_len;
}

pub fn parseByte(text: []const u8) !u8 {
    return std.fmt.parseUnsigned(u8, text, 16);
}

pub const fdPayloadLengthFromDlc = trace_dlc.fdPayloadLengthFromDlc;

test "parses TRC millisecond timestamps into nanoseconds" {
    try std.testing.expectEqual(@as(u64, 0), try parseTimestampMsToNs("0"));
    try std.testing.expectEqual(@as(u64, 1_234_000), try parseTimestampMsToNs("1.234"));
    try std.testing.expectEqual(@as(u64, 1_234_567), try parseTimestampMsToNs("1.234567"));
    try std.testing.expectError(error.InvalidTimestamp, parseTimestampMsToNs("-1.0"));
}

test "parses TRC ID width into standard or extended IDs" {
    try std.testing.expectEqual(Id{ .value = 0x123, .is_extended = false }, try idFromText("0123"));
    try std.testing.expectEqual(Id{ .value = 0x18fee900, .is_extended = true }, try idFromText("18FEE900"));
    try std.testing.expectError(error.InvalidId, idFromText("0800"));
}

File: wasm/src/trc/frame.rs
use std::num::IntErrorKind;

use crate::trace::{
    CanId, ExtraPrecision, TraceError, decimal_fraction_to_units, fd_payload_length_from_dlc,
};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(super) enum Version {
    V10,
    V11,
    V12,
    V13,
    V20,
    V21,
    V30,
}

impl Version {
    pub(super) fn from_text(text: &str) -> Result<Self, TraceError> {
        match text {
            "1.0" => Ok(Self::V10),
            "1.1" => Ok(Self::V11),
            "1.2" => Ok(Self::V12),
            "1.3" => Ok(Self::V13),
            "2.0" => Ok(Self::V20),
            "2.1" => Ok(Self::V21),
            "3.0" => Ok(Self::V30),
            _ => Err(TraceError::UnsupportedTrcVersion),
        }
    }

    pub(super) const fn is_v2(self) -> bool {
        matches!(self, Self::V20 | Self::V21)
    }
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub(super) struct ColumnMap {
    pub(super) number: Option<usize>,
    pub(super) offset: Option<usize>,
    pub(super) record_type: Option<usize>,
    pub(super) bus: Option<usize>,
    pub(super) id: Option<usize>,
    pub(super) direction: Option<usize>,
    pub(super) reserved: Option<usize>,
    pub(super) data_len: Option<usize>,
    pub(super) dlc: Option<usize>,
    pub(super) data: Option<usize>,
    pub(super) token_count_before_data: usize,
}

impl ColumnMap {
    pub(super) fn from_text(text: &str) -> Result<Self, TraceError> {
        let mut map = Self::default();
        for (index, raw_part) in text.split(',').enumerate() {
            match raw_part.trim_matches([' ', '\t', '\r']) {
                "N" => map.number = Some(index),
                "O" => map.offset = Some(index),
                "T" => map.record_type = Some(index),
                "B" => map.bus = Some(index),
                "I" => map.id = Some(index),
                "d" => map.direction = Some(index),
                "R" => map.reserved = Some(index),
                "l" => map.data_len = Some(index),
                "L" => map.dlc = Some(index),
                "D" => map.data = Some(index),
                _ => return Err(TraceError::UnsupportedTrcColumn),
            }
        }

        if map.offset.is_none()
            || map.record_type.is_none()
            || map.id.is_none()
            || map.direction.is_none()
            || map.data.is_none()
            || (map.data_len.is_none() && map.dlc.is_none())
        {
            return Err(TraceError::InvalidTrcColumns);
        }
        map.token_count_before_data = map.data.expect("data column checked above");
        Ok(map)
    }
}

pub(crate) fn parse_timestamp_ms_to_ns(text: &str) -> Result<u64, TraceError> {
    if text.is_empty() || text.starts_with('-') {
        return Err(TraceError::InvalidTimestamp);
    }
    let mut parts = text.split('.');
    let milliseconds = parts
        .next()
        .ok_or(TraceError::InvalidTimestamp)?
        .parse::<u64>()
        .map_err(|error| match error.kind() {
            IntErrorKind::PosOverflow => TraceError::TimestampOverflow,
            _ => TraceError::InvalidTimestamp,
        })?;
    let fraction = parts.next();
    if parts.next().is_some() {
        return Err(TraceError::InvalidTimestamp);
    }

    let mut nanoseconds = milliseconds
        .checked_mul(1_000_000)
        .ok_or(TraceError::TimestampOverflow)?;
    if let Some(fraction) = fraction {
        nanoseconds = nanoseconds
            .checked_add(decimal_fraction_to_units(
                fraction,
                1_000_000,
                6,
                ExtraPrecision::Reject,
            )?)
            .ok_or(TraceError::TimestampOverflow)?;
    }
    Ok(nanoseconds)
}

pub(super) fn id_from_text(text: &str) -> Result<CanId, TraceError> {
    if !matches!(text.len(), 4 | 8) {
        return Err(TraceError::InvalidId);
    }
    let value = u32::from_str_radix(text, 16).map_err(|_| TraceError::InvalidId)?;
    if text.len() == 8 {
        CanId::extended(value)
    } else {
        CanId::standard(value)
    }
}

pub(super) fn parse_dlc(text: &str) -> Result<u8, TraceError> {
    let dlc = text.parse::<u8>().map_err(|_| TraceError::InvalidDlc)?;
    (dlc <= 15).then_some(dlc).ok_or(TraceError::InvalidDlc)
}

pub(super) fn parse_payload_length(text: &str) -> Result<u8, TraceError> {
    let length = text
        .parse::<u8>()
        .map_err(|_| TraceError::InvalidPayloadLength)?;
    (length <= 64)
        .then_some(length)
        .ok_or(TraceError::InvalidPayloadLength)
}

pub(super) fn parse_byte(text: &str) -> Result<u8, TraceError> {
    u8::from_str_radix(text, 16).map_err(|_| TraceError::InvalidFrameLine)
}

pub(super) fn fd_length(dlc: u8) -> Result<u8, TraceError> {
    fd_payload_length_from_dlc(dlc)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_millisecond_timestamps_without_floats() {
        assert_eq!(parse_timestamp_ms_to_ns("0"), Ok(0));
        assert_eq!(parse_timestamp_ms_to_ns("1.234"), Ok(1_234_000));
        assert_eq!(parse_timestamp_ms_to_ns("1.234567"), Ok(1_234_567));
        assert_eq!(
            parse_timestamp_ms_to_ns("-1.0"),
            Err(TraceError::InvalidTimestamp)
        );
    }

    #[test]
    fn parses_id_width_into_standard_or_extended_identity() {
        assert_eq!(id_from_text("0123"), CanId::standard(0x123));
        assert_eq!(id_from_text("18FEE900"), CanId::extended(0x18fee900));
        assert_eq!(id_from_text("0800"), Err(TraceError::InvalidId));
    }

    #[test]
    fn validates_column_contract() {
        let columns = ColumnMap::from_text("N,O,T,B,I,d,R,L,D").unwrap();
        assert_eq!(columns.offset, Some(1));
        assert_eq!(columns.data, Some(8));
        assert_eq!(
            ColumnMap::from_text("N,O"),
            Err(TraceError::InvalidTrcColumns)
        );
    }
}
